### DISCUSSION TITLE
エッジランタイムではStaticなAPI Routeを作成するのは無理なのか？

### DISCUSSION THREAD

**Discussion by kagomen** [lgtm-factory/lgtm-factory#194] [Q&A] [answered]
@kazzyfrog 

#185 の実装に関して、質問です！

`generateStaticParams()`を使用してSG化にチャレンジしているのですが、エラーが出てしまいます

というのも、API Routeを使用するときはSSG RouteでもStatic Routeでも、`export const runtime = "edge";`が必要っぽくて、そうなるとビルドした時にDynamic Routeとして生成されるそうなんです😭（以下、参考画像）

### Static時
<img width="687" alt="スクリーンショット 2024-09-07 23 09 40" src="https://github.com/user-attachments/assets/f47391b2-0304-4c83-a420-a9f5870589d0">

### SSG時
<img width="691" alt="スクリーンショット 2024-09-07 23 10 11" src="https://github.com/user-attachments/assets/b7e08187-f115-4f3b-bea1-71a22a870426">

### edge指定したら実行はできたけどDynamic routeになった
（fsモジュールは使えないので削除）
<img width="417" alt="スクリーンショット 2024-09-07 23 21 56" src="https://github.com/user-attachments/assets/e90c2be7-1949-4b0f-b88d-a3912bdcc8e5">

`importLgtmDataModule()`の動的importが原因かな？と思って、静的importでひとつのテーマだけ仮にimportして、`npm run preview`を試してみたのですが変わらずで、`export const runtime = "edge";`が必要ですーって出ました

エッジ環境ではStaticなAPIルートを作成するのは無理なのか、それとも自分が根本的な実装の間違いをしているのか、判断ができず困っています😿
お手隙の時にアドバイスいただけると幸いです……🙏

### 実装したソースコード

```ts
// api/v1/design-info/[theme]/route.tsx

import { NextRequest, NextResponse } from "next/server";
import { importLgtmDataModule } from "@/app/api/v1/lgtm-images/_utils";
import fs from "fs/promises";
import path from "path";

export const dynamic = "force-static";  // これを無くすとSSG Routeになる

async function getFileNames() {
  const fileNames = await fs.readdir("lgtm-data");
  const themes = fileNames
    .map((file) => path.parse(file).name)
    .filter((theme) => theme !== "error");
  return themes;
}

export async function generateStaticParams() {
  const themes = await getFileNames();
  const themesList = themes.map((theme) => ({
    theme: theme,
  }));
  await getFileNames();
  return [themesList];
}

export async function GET(
  _: NextRequest,
  { params }: { params: { theme: string } },
) {
  const { theme } = params;

  if (!theme) {
    return NextResponse.json({ designInfo: null }, { status: 400 });
  }

  const inputData = {
    emoji: "",
    text: "",
    color: "",
  };

  try {
    const getLgtmData = await importLgtmDataModule(theme);
    const { designInfo } = await getLgtmData(inputData);
    return NextResponse.json({ designInfo: designInfo }, { status: 200 });
  } catch (_) {
    return NextResponse.json({ designInfo: null }, { status: 400 });
  }
}
```

**Comment by kazzyfrog** [comment]
ビルド`npm run build`は成功しますかね？

**Reply by kagomen** [reply]
`npm run build`と`npm run start`だと、ばっちり動いてくれます！
<img width="585" alt="スクリーンショット 2024-09-08 0 10 41" src="https://github.com/user-attachments/assets/0b9ffb46-8fe8-4e96-a6ec-808536932880">

**Comment by kazzyfrog** [comment]
ちなみに、下記を追加しても無理ですよね。。。？
```
export const dynamicParams = false
```

**Reply by kagomen** [reply]
やってみたところ、

- `npm run preview`: 同じエラーが発生
- `npm run build`->`npm run start`:  ビルドは成功、任意のdesin-info/[theme]にアクセスすると404 Not Foundページに

という感じになりました😿
自分もGitHubのイシューなど見てみます！

**Comment by kagomen** [comment]
https://developers.cloudflare.com/pages/framework-guides/nextjs/ssr/troubleshooting/#generatestaticparams

> generateStaticParams
When you use [static site generation (SSG) ↗](https://nextjs.org/docs/pages/building-your-application/rendering/static-site-generation) in the [/app directory ↗](https://nextjs.org/docs/getting-started/project-structure) and also use the [generateStaticParams ↗](https://nextjs.org/docs/app/api-reference/functions/generate-static-params) function, Next.js tries to handle requests for non statically generated routes automatically, and creates a [Node.js runtime serverless function ↗](https://nextjs.org/docs/app/building-your-application/rendering/edge-and-nodejs-runtimes) that is not compatible with Cloudflare Pages.

You can opt out of this behavior by setting [dynamicParams ↗](https://nextjs.org/docs/app/api-reference/file-conventions/route-segment-config#dynamicparams) to false:

これで解決するはずなのに、なぜしないのか

---

https://nextjs.org/docs/app/api-reference/file-conventions/route-segment-config#dynamicparams

> export const dynamicParams = true // true | false,
true (default): Dynamic segments not included in generateStaticParams are generated on demand.
false: Dynamic segments not included in generateStaticParams will return a 404.

404になるのは正しい挙動

**Reply by kagomen** [reply]
```
^._ .^= ∫ lgtm-factory % npm run preview

> lgtm-factory@0.1.0 preview
> npm run pages:build && wrangler pages dev


> lgtm-factory@0.1.0 pages:build
> npx @cloudflare/next-on-pages

⚡️ @cloudflare/next-on-pages CLI v.1.13.2
⚡️ Detected Package Manager: npm (10.8.1)
⚡️ Preparing project...
⚡️ Project is ready
⚡️ Building project...
▲  Vercel CLI 35.2.4
▲  WARNING: You should not upload the `.next` directory.
▲  Installing dependencies...
▲  > lgtm-factory@0.1.0 prepare
▲  > husky
▲  up to date in 496ms
▲  177 packages are looking for funding
▲  run `npm fund` for details
▲  Detected Next.js version: 14.2.5
▲  Detected `package-lock.json` generated by npm 7+
▲  Running "npm run build"
▲  > lgtm-factory@0.1.0 build
▲  > next build
▲  ▲ Next.js 14.2.5
▲  - Environments: .env.production
▲  Creating an optimized production build ...
▲  ✓ Compiled successfully
▲  Linting and checking validity of types ...
▲  Collecting page data ...
▲  ⚠ Using edge runtime on a page currently disables static generation for that page
▲  Generating static pages (0/9) ...
▲  Generating static pages (2/9)
▲  Generating static pages (4/9)
▲  Generating static pages (6/9)
▲  ✓ Generating static pages (9/9)
▲  Finalizing page optimization ...
▲  Collecting build traces ...
▲  
▲  Route (app)                              Size     First Load JS
▲  ┌ ○ /                                    5.71 kB         113 kB
▲  ├ ○ /_not-found                          144 B          87.2 kB
▲  ├ ƒ /api/v1/design-info                  0 B                0 B
▲  ├ ● /api/v1/design-info/[theme]          0 B                0 B
▲  ├ ƒ /api/v1/lgtm-images                  0 B                0 B
▲  ├ ○ /apple-icon.png                      0 B                0 B
▲  ├ ○ /icon.svg                            0 B                0 B
▲  ├ ○ /manifest.json                       0 B                0 B
▲  └ ○ /terms                               144 B          87.2 kB
▲  + First Load JS shared by all            87.1 kB
▲  ├ chunks/23-547baf2c39fe4cc5.js        31.5 kB
▲  ├ chunks/fd9d1056-c4050f5b61ad0a3b.js  53.6 kB
▲  └ other shared chunks (total)          1.92 kB
▲  ○  (Static)   prerendered as static content
▲  ●  (SSG)      prerendered as static HTML (uses getStaticProps)
▲  ƒ  (Dynamic)  server-rendered on demand
▲  Traced Next.js server files in: 104.942ms
▲  Created all serverless functions in: 33.615ms
▲  Collected static files (public/, static/, .next/static): 12.09ms
▲  Build Completed in .vercel/output [13s]
⚡️ Completed `npx vercel build`.

⚡️ ERROR: Failed to produce a Cloudflare Pages build from the project.
⚡️ 
⚡️      The following routes were not configured to run with the Edge Runtime:
⚡️        - /api/v1/design-info/[theme]
⚡️ 
⚡️      Please make sure that all your non-static routes export the following edge runtime route segment config:
⚡️        export const runtime = 'edge';
⚡️ 
⚡️      You can read more about the Edge Runtime on the Next.js documentation:
⚡️        https://nextjs.org/docs/app/building-your-application/rendering/edge-and-nodejs-runtimes
```

**Reply by kagomen** [reply]
https://github.com/cloudflare/next-on-pages/issues/745

> 私の場合、1ヶ月前はdynamicParams = false;とgenerateStaticParamsのおかげで動作していましたが、今日同じコードを使用すると同じエラーが発生します： 次のルートは、Edgeランタイムで実行するように設定されていません。 next-on-pagesはNextJSの動作が非常に不安定で、プロジェクトを公開しようとするたびに、どこからともなく湧いてくる新しい問題を修正しなければならないような気がします。

修正PRマージされたけど、直っていないっぽい？同じ人たくさんいる
あとkazzyfrogさんもいた！報告ありがとうございます！！

こういうのって「私も同じです！」って言った方が、Cloudflareが対応してくれる確率上がるんだろうか・・・

**Reply by kazzyfrog** [reply]
メンテナーが反応してくれることは結構ありますよ！
それ以外にも、コメントしている人や、それを見た開発者が、別の１時的な解決策で対処した場合に、
それを教えてくれたりもします👍

そもそも、対処されているっぽいので、、自分に原因があるのかどうかもわかっていないので！

**Reply by kagomen** [reply]
なるほどです！
明日以降、スレッドが下がってたら上げ直しの意味も込めて書いてみようかな……

**Comment by kazzyfrog** [comment]
もし可能だったら、useEffectでデザイン情報を取得している処理を、
`src/components/MainSection.tsx `内で、サーバーサイドで`fetch()`する形にする修正でも、一旦OKかもです！

現状、毎回ページが描画されるごとに、すべてのデザイン情報が取得されるので、
`fetch`でキャッシュする形がいいかなと！

**Reply by kazzyfrog** [reply]
APIのルートは、動的のままで、サーバーサイドのfetchでキャッシュする形ですかね！

**Reply by kazzyfrog** [reply]
あ、`src/components/ImageInfoModal.tsx`で、`fetch()`する形でも、問題なさそう！
`use client`をなくす形で！

**Reply by kagomen** [reply]
> APIのルートは、動的のままで、サーバーサイドのfetchでキャッシュする形

そうします！👍

それから、api/v1/design-info/[theme]/route.tsxに書いた内容もPRに含めても大丈夫ですか？
SSG Routeが使えるようになったら、そのまま動かせると思うので！🙌

**Reply by kazzyfrog** [reply]
確かにですね！！

そしたら、`api/v1/design-info/[theme]/route.tsx`をエッジランタイムで、動的にして、
そこから`fetch`すれば、問題なくいけそうですかね？

現在ある、デザイン情報のAPIは削除する形でいいかと🙌

**Reply by kagomen** [reply]
edge使うので、`fs`の処理を変える感じになりますね
ちょっと考えてみます！👍

**Reply by kazzyfrog** [reply]
あ、確かにそうですね！

そしたら、`api/v1/design-info/[theme]/route.tsx`含めなくてもいいかもです！

それと、これ含めてもCloudflareのビルドエラーにならない感じですかね？

**Reply by kagomen** [reply]
> fsの処理を変える

言ってから気付きましたが、無理かもです💧

> これ含めてもCloudflareのビルドエラーにならない感じですかね？

ビルドエラーになるので、txtファイルとしてあげておこうかなーって感じでした！

**Reply by kazzyfrog** [reply]
それでしたら、
このスレッド、もしくは新しいスレッドに、コードをコピーして、
そのスレの`#`番号をPRでリンクするとかの方が、良いかもです👍

**Reply by kagomen** [reply]
> このスレッド、もしくは新しいスレッドに、コードをコピーして、
その#番号をPRでリンクするとかの方が、良いかもです👍

了解です！
そうしますー！👍

**Reply by kagomen** [reply]
@kazzyfrog 
一応ソース完成したんですが、sisoさんのPRとコンフリクト起こすと思うので、あちらがマージされてからPR出します🙌

ちなみになんですが、
sisoさんのPRで、自分が提案してる処理がuseState使ったものなので、それが採用されればImageInfoModalからInput以下を切り出す形になるかと思うのですが、大丈夫ですかね・・・？🤔

それか、useState使わないもっとうまいやり方がありそうな気もしているのですが、どうなんでしょう・・・？

**Reply by kagomen** [reply]
MainSectionで
```ts
fetch(
    `${siteMetadata.SITE_URL}/api/v1/design-info?theme=${theme}`,
  );
```
ができればコンフリクトによる修正箇所も少なくて良いなあと思うのですが、mapで回してる`theme`をfetchの処理と繋げる方法が、自分じゃ思いつかなくて・・・😭

**Reply by kazzyfrog** [reply]
> 一応ソース完成したんですが、sisoさんのPRとコンフリクト起こすと思うので、あちらがマージされてからPR出します🙌

了解です👍

ImageInfoModalから、`SheetContent`内のフォームの部分を、切り出してもいいかもですね！
その中に、画像も含める感じにすれば、Sheetにも依存せず、いい感じに切り分けられそう？かもです！

簡潔にしたら、こんなイメージ？でいけるかな。。
```jsx
<Sheet>
       <SheetTrigger>
         <LgtmImage url={url} className="cursor-pointer" />
       </SheetTrigger>
       <SheetContent className="space-y-8 p-10">
         <SheetHeader>
         <ImageFrom> // ここに、モーダル内で表示する画像も含める
       </SheetContent>
</Sheet>
```

**Reply by kagomen** [reply]
さっとやってみた感じ、うまくいきそうでした！👍👍
こんな風に切り分けたら良いんですね🔍
ありがとうございますー！🙌

**Comment by kagomen** [comment, marked as answer]
以下の処理は2024.9.8現在、ビルドエラーになり使えない状態です
Vercelにデプロイするか、`dynamicParams`フラグが使えるようになった際に、以下のパスにroute.tsxを作成し、ソースコードをコピペして使用してください
パス：/api/v1/design-info/[theme]

```ts
import { NextRequest, NextResponse } from "next/server";
import { importLgtmDataModule } from "@/app/api/v1/lgtm-images/_utils";
import fs from "fs/promises";
import path from "path";

// CloudflareでSSG Routeを使う際は以下が必要
// export const dynamicParams = false;

// Next.jsでStatic Routeを使う際は以下が必要
// export const dynamic = 'force-static'

async function getFileNames() {
  const fileNames = await fs.readdir("lgtm-data");
  const themes = fileNames
    .map((file) => path.parse(file).name)
    .filter((theme) => theme !== "error");
  return themes;
}

export async function generateStaticParams() {
  const themes = await getFileNames();
  const themesList = themes.map((theme) => ({
    theme: theme,
  }));
  await getFileNames();
  return [themesList];
}

export async function GET(
  _: NextRequest,
  { params }: { params: { theme: string } },
) {
  const { theme } = params;

  if (!theme) {
    return NextResponse.json({ designInfo: null }, { status: 400 });
  }

  const inputData = {
    emoji: "",
    text: "",
    color: "",
  };

  try {
    const getLgtmData = await importLgtmDataModule(theme);
    const { designInfo } = await getLgtmData(inputData);
    return NextResponse.json({ designInfo: designInfo }, { status: 200 });
  } catch (_) {
    return NextResponse.json({ designInfo: null }, { status: 400 });
  }
}
```
